piet-cairo/src/text.rs

```diff
@@ -1,64 +1,162 @@
+    //Calculated on build
+    size: Size,
+    ink_size: Size,
+    pango_offset: Vec2,
+    trailing_ws_width: f64,
     pub(crate) line_metrics: Vec<LineMetric>,
+    pub(crate) pango_layout: PangoLayout,
 }
 
 pub struct CairoTextLayoutBuilder {
     text: Rc<dyn TextStorage>,
     defaults: util::LayoutDefaults,
+    attributes: Vec<AttributeWithRange>,
+    last_range_start_pos: usize,
     width_constraint: f64,
+    pango_layout: PangoLayout,
+}
+
+struct AttributeWithRange {
+    attribute: TextAttribute,
+    range: Option<Range<usize>>, //No range == entire layout
+}
+
+impl AttributeWithRange {
+    fn into_pango(self) -> Option<PangoAttribute> {
+        let mut pango_attribute = match &self.attribute {
+            TextAttribute::FontFamily(family) => {
+                let family = family.name();
+                /*
+                 * NOTE: If the family fails to resolve we just don't apply the attribute.
+                 * That allows Pango to use its default font of choice to render that text
+                 */
+                PangoAttribute::new_family(family)?
+            }
+
+            TextAttribute::FontSize(size) => {
+                let size = (size * PANGO_SCALE) as i32;
+                PangoAttribute::new_size_absolute(size).unwrap()
+            }
+
+            TextAttribute::Weight(weight) => {
+                //This is horrid
+                let pango_weights = [
+                    (100, PangoWeight::Thin),
+                    (200, PangoWeight::Ultralight),
+                    (300, PangoWeight::Light),
+                    (350, PangoWeight::Semilight),
+                    (380, PangoWeight::Book),
+                    (400, PangoWeight::Normal),
+                    (500, PangoWeight::Medium),
+                    (600, PangoWeight::Semibold),
+                    (700, PangoWeight::Bold),
+                    (800, PangoWeight::Ultrabold),
+                    (900, PangoWeight::Heavy),
+                    (1_000, PangoWeight::Ultraheavy),
+                ];
+
+                let weight = weight.to_raw() as i32;
+                let mut closest_index = 0;
+                let mut closest_distance = 2_000; //Random very large value
+                for (current_index, pango_weight) in pango_weights.iter().enumerate() {
+                    let distance = (pango_weight.0 - weight).abs();
+                    if distance < closest_distance {
+                        closest_distance = distance;
+                        closest_index = current_index;
+                    }
+                }
+
+                PangoAttribute::new_weight(pango_weights[closest_index].1).unwrap()
+            }
+
+            TextAttribute::TextColor(text_color) => {
+                let (r, g, b, _) = text_color.as_rgba();
+                PangoAttribute::new_foreground(
+                    (r * 65535.) as u16,
+                    (g * 65535.) as u16,
+                    (b * 65535.) as u16,
+                )
```

Comment:
I'd skip the f64 conversion and just do `text_color.as_rgba8` and then `(r as u16 * 256) + (r as u16)`

Reply:
I honestly don't remember if I thought it up or if it was derived from some earlier code. Either way that looks more reasonable